src/ints/mouse_dos_driver.cpp

```diff
@@ -0,0 +1,1440 @@
+    case 0x14: // Exchange event-handler
+        if (reg_bx != 0)
+            SegSet16(es, reg_bx);
+        else
+            SegSet16(es, SegValue(ds));
+        break;
+    case 0x10: // Define screen region for updating
+        reg_cx = real_readw(SegValue(ds), rdxpt);
+        reg_dx = real_readw(SegValue(ds), static_cast<uint16_t>(rdxpt + 2));
+        reg_si = real_readw(SegValue(ds), static_cast<uint16_t>(rdxpt + 4));
+        reg_di = real_readw(SegValue(ds), static_cast<uint16_t>(rdxpt + 6));
+        break;
+    default: break;
+    }
+
+    INT33_Handler();
+
+    // save back the registers, too
+    real_writew(SegValue(ds), raxpt, reg_ax);
+    real_writew(SegValue(ds), rbxpt, reg_bx);
+    real_writew(SegValue(ds), rcxpt, reg_cx);
+    real_writew(SegValue(ds), rdxpt, reg_dx);
+    switch (rax) {
+    case 0x1f: // Disable Mousedriver
+        real_writew(SegValue(ds), rbxpt, SegValue(es));
+        break;
+    case 0x14: // Exchange event-handler
+        real_writew(SegValue(ds), rcxpt, SegValue(es));
+        break;
+    default: break;
+    }
+
+    return CBRET_NONE;
+}
+
+uintptr_t UIR_Handler()
+{
+    mouse_shared.dos_cb_running = false;
+    return CBRET_NONE;
+}
+
+bool MOUSEDOS_HasCallback(const uint8_t mask)
+{
+    return state.sub_mask & mask;
+}
+
+uintptr_t MOUSEDOS_DoCallback(const uint8_t mask, const MouseButtons12S buttons_12S)
+{
+    mouse_shared.dos_cb_running = true;
+    const bool is_wheel = mask & static_cast<uint8_t>(MouseEventId::WheelHasMoved);
+
+    reg_ax = mask;
+    reg_bl = buttons_12S.data;
+    reg_bh = is_wheel ? GetResetWheel8bit() : 0;
+    reg_cx = GetPosX();
+    reg_dx = GetPosY();
+    reg_si = SignedToReg16(state.mickey_x);
+    reg_di = SignedToReg16(state.mickey_y);
+
+    CPU_Push16(RealSeg(uir_callback));
+    CPU_Push16(RealOff(uir_callback));
+    CPU_Push16(state.sub_seg);
+    CPU_Push16(state.sub_ofs);
+
+    return CBRET_NONE;
+}
+
+void MOUSEDOS_Init()
+{
+    // Callback for mouse interrupt 0x33
+    auto call_int33 = CALLBACK_Allocate();
+    // RealPt i33loc = RealMake(CB_SEG + 1,(call_int33 * CB_SIZE) - 0x10);
+    RealPt i33loc = RealMake(static_cast<uint16_t>(DOS_GetMemory(0x1) - 1), 0x10);
+    CALLBACK_Setup(call_int33, &INT33_Handler, CB_MOUSE, Real2Phys(i33loc), "Mouse");
+    // Wasteland needs low(seg(int33))!=0 and low(ofs(int33))!=0
+    real_writed(0, 0x33 << 2, i33loc);
+
+    auto call_mouse_bd = CALLBACK_Allocate();
+    CALLBACK_Setup(call_mouse_bd,
+                   &MOUSE_BD_Handler,
+                   CB_RETF8,
+                   PhysMake(RealSeg(i33loc), static_cast<uint16_t>(RealOff(i33loc) + 2)),
+                   "MouseBD");
+    // pseudocode for CB_MOUSE (including the special backdoor entry point):
+    //    jump near i33hd
+    //    callback MOUSE_BD_Handler
+    //    retf 8
+    //  label i33hd:
+    //    callback INT33_Handler
+    //    iret
+
+    // Callback for mouse user routine return
+    auto call_uir = CALLBACK_Allocate();
```

Comment:
call_user_irq_routine ? (i'm guessing)

Reply:
User callback actually. Changed name to something less cryptic.